README.md
<img align="right" src="https://raw.github.com/cliffano/sendman/master/avatar.jpg" alt="Avatar"/>

[![Build Status](https://img.shields.io/travis/cliffano/sendman.svg)](http://travis-ci.org/cliffano/sendman)
[![Dependencies Status](https://img.shields.io/david/cliffano/sendman.svg)](http://david-dm.org/cliffano/sendman)
[![Coverage Status](https://img.shields.io/coveralls/cliffano/sendman.svg)](https://coveralls.io/r/cliffano/sendman?branch=master)
[![Published Version](https://img.shields.io/npm/v/sendman.svg)](http://www.npmjs.com/package/sendman)
<br/>
[![npm Badge](https://nodei.co/npm/sendman.png)](http://npmjs.org/package/sendman)

Sendman
-------

Sendman is a CLI tool for sending files to a local/remote location.

This is handy when you want to use a configuration file to define the local/remote server host, port, username, password, local path, and remote path of the files you want to send via FTP, SCP, rsync, or AWS S3. No more retyping lengthy commands one by one.

Installation
------------

    npm install -g sendman

Usage
-----

Create a sample .sendman.json configuration file:

    sendman init

Configuration file contains:

    {
      "protocol": "scp",
      "host": "somehost",
      "port": 22,
      "username": "someusername",
      "password": "somepassword",
      "local": "/path/to/local/dir",
      "remote": "/path/to/remote/dir"
    }

Send the files located at `local` path to `remote` location:

    sendman send

It's possible to include sub-configuration file.

.sendman.json contains:

    {
      "protocol": "ftp",
      "host": "somehost",
      "port": 21,
      "local": "/path/to/local/dir",
      "remote": "/path/to/remote/dir",
      "parallel": 10,
      "include": "/path/to/.secretSendman.json"
    }

.secretSendman.json contains:

    {
      "username": "someusername",
      "password": "somepassword"
    }

Alternatively, you can use rsync to send files to a remote host.

    {
      "protocol": "rsync",
      "host": "somehost",
      "local": "/path/to/local/dir",
      "remote": "/path/to/remote/dir"
    }

rsync can also be used to send files between local paths.

    {
      "protocol": "rsync",
      "source": "/path/to/local/dir1",
      "destination": "/path/to/local/dir2"
    }

To send files to an AWS S3 bucket:

    {
      "protocol": "s3",
      "local": "/path/to/local/dir",
      "bucket": "somebucket",
      "region": "ap-southeast-2",
      "remote": "/path/to/bucket/prefix",
      "accessKeyId": "someaccesskeyid",
      "secretAccessKey": "somesecretaccesskey"
    }

Colophon
--------

[Developer's Guide](http://cliffano.github.io/developers_guide.html#nodejs)

Build reports:

* [Code complexity report](http://cliffano.github.io/sendman/complexity/plato/index.html)
* [Unit tests report](http://cliffano.github.io/sendman/test/buster.out)
* [Test coverage report](http://cliffano.github.io/sendman/coverage/buster-istanbul/lcov-report/lib/index.html)
* [Integration tests report](http://cliffano.github.io/sendman/test-integration/cmdt.out)
* [API Documentation](http://cliffano.github.io/sendman/doc/dox-foundation/index.html)


CHANGELOG.md
### 1.0.2
*

### 1.0.1
* Set min node engine to >= 4.0.0

### 1.0.0
* Major version upgrade only

### 0.2.2
* Add s3 protocol support
* Update ftp protocol output following data structure change in ftp-deploy v1.1.0

### 0.2.1
* Add ftp error handling

### 0.2.0
* Rename 'file' property to 'include'

### 0.1.0
* Add rsync protocol support
* Set min node engine to >= v0.10.0

### 0.0.3
* Add SCP protocol support

### 0.0.2
* Change display output to only display uploaded files with path and count

### 0.0.1
* Initial version


test/cli.js
var buster = require('buster-node');
var _cli = require('bagofcli');
var cli = require('../lib/cli');
var referee = require('referee');
var Sendman = require('../lib/sendman');
var assert = referee.assert;

buster.testCase('cli - exec', {
  'should contain commands with actions': function (done) {
    var mockCommand = function (base, actions) {
      assert.defined(base);
      assert.defined(actions.commands.init.action);
      assert.defined(actions.commands.send.action);
      done();
    };
    this.mock({});
    this.stub(_cli, 'command', mockCommand);
    cli.exec();
  }
});

buster.testCase('cli - init', {
  setUp: function () {
    this.mockConsole = this.mock(console);
  },
  'should contain init command and delegate to Sendman init when exec is called': function (done) {
    this.mockConsole.expects('log').withExactArgs('Creating sample configuration file: .sendman.json');
    this.stub(_cli, 'command', function (base, actions) {
      actions.commands.init.action();
    });
    this.stub(Sendman.prototype, 'init', function (cb) {
      assert.equals(typeof cb, 'function');
      done();
    });
    cli.exec();
  }
});

buster.testCase('cli - send', {
  setUp: function () {
    this.mockProcess = this.mock(process);
  },
  'should use custom config file when arg is provided': function () {
    this.mockProcess.expects('exit').once().withExactArgs(0);
    this.stub(_cli, 'command', function (base, actions) {
      actions.commands.send.action({
        file: '.somesendman.json'
      });
    });
    this.stub(Sendman.prototype, 'send', function (file, cb) {
      assert.equals(file, '.somesendman.json');
      cb();
    });
    cli.exec();
  },
  'should use default config file when arg is not provided': function () {
    this.mockProcess.expects('exit').once().withExactArgs(0);
    this.stub(_cli, 'command', function (base, actions) {
      actions.commands.send.action();
    });
    this.stub(Sendman.prototype, 'send', function (file, cb) {
      assert.equals(file, '.sendman.json');
      cb();
    });
    cli.exec();
  }
});


test/sendman.js
var buster = require('buster-node');
var cli = require('bagofcli');
var fsx = require('fs.extra');
var FtpDeploy = require('ftp-deploy');
var proxyquire = require('proxyquire');
var referee = require('referee');
var rsyncwrapper = require('rsyncwrapper');
var s3 = require('s3');
var scp2 = require('scp2');
var Sendman = require('../lib/sendman');
var assert = referee.assert;

buster.testCase('sendman - init', {
  'should copy sample couchpenter.js file to current directory when init is called': function (done) {
    this.mock({});
    this.stub(fsx, 'copy', function (src, dest, cb) {
      assert.isTrue(src.match(/\/examples\/.sendman.json$/).length === 1);
      assert.equals(dest, '.sendman.json');
      cb();
    });
    var sendman = new Sendman();
    sendman.init(function (err, result) {
      assert.equals(err, undefined);
      done();
    });
  }
});

buster.testCase('sendman - send', {
  setUp: function () {
    this.mockCli = this.mock(cli);
  },
  'should pass error when protocol is unsupported': function (done) {
    var data = { protocol: 'inexistingprotocol' };
    this.mockCli.expects('lookupFile').once().withExactArgs('.sendman.json').returns(JSON.stringify(data));
    var sendman = new Sendman();
    sendman.send('.sendman.json', function (err) {
      assert.equals(err.message, 'Unsupported protocol %s', 'inexistingprotocol');
      done();
    });
  },
  'should merge sub-configuration file when specified in main configuration file': function (done) {
    var data = { protocol: 'ftp', include: '.sub.sendman.json' };
    var subData = { foo: 'bar' };
    this.mockCli.expects('lookupFile').once().withExactArgs('.sendman.json').returns(JSON.stringify(data));
    this.mockCli.expects('lookupFile').once().withExactArgs('.sub.sendman.json').returns(JSON.stringify(subData));
    var sendman = new Sendman();
    sendman._ftp = function (opts, cb) {
      assert.equals(opts.protocol, 'ftp');
      assert.equals(opts.foo, 'bar');
      cb();
    };
    sendman.send('.sendman.json', function (err) {
      assert.equals(err, undefined);
      done();
    });
  },
  'should send file via ftp when protocol is ftp': function (done) {
    var data = { protocol: 'ftp' };
    this.mockCli.expects('lookupFile').once().withExactArgs('.sendman.json').returns(JSON.stringify(data));
    var sendman = new Sendman();
    sendman._ftp = function (opts, cb) {
      assert.equals(opts.protocol, 'ftp');
      cb();
    };
    sendman.send('.sendman.json', function (err) {
      assert.equals(err, undefined);
      done();
    });
  },
  'should send file via AWS S3 when protocol is s3': function (done) {
    var data = { protocol: 's3' };
    this.mockCli.expects('lookupFile').once().withExactArgs('.sendman.json').returns(JSON.stringify(data));
    var sendman = new Sendman();
    sendman._s3 = function (opts, cb) {
      assert.equals(opts.protocol, 's3');
      cb();
    };
    sendman.send('.sendman.json', function (err) {
      assert.equals(err, undefined);
      done();
    });
  },
  'should send file via scp when protocol is scp': function (done) {
    var data = { protocol: 'scp' };
    this.mockCli.expects('lookupFile').once().withExactArgs('.sendman.json').returns(JSON.stringify(data));
    var sendman = new Sendman();
    sendman._scp = function (opts, cb) {
      assert.equals(opts.protocol, 'scp');
      cb();
    };
    sendman.send('.sendman.json', function (err) {
      assert.equals(err, undefined);
      done();
    });
  },
  'should send file via rsync when protocol is rsync': function (done) {
    var data = { protocol: 'rsync' };
    this.mockCli.expects('lookupFile').once().withExactArgs('.sendman.json').returns(JSON.stringify(data));
    var sendman = new Sendman();
    sendman._rsync = function (opts, cb) {
      assert.equals(opts.protocol, 'rsync');
      cb();
    };
    sendman.send('.sendman.json', function (err) {
      assert.equals(err, undefined);
      done();
    });
  }
});

buster.testCase('sendman - ftp', {
  setUp: function () {
    this.mockConsole = this.mock(console);
  },
  'should set opts for ftp module': function (done) {
    this.mockConsole.expects('log').once().withExactArgs('%s (%d/%d)', 'some/path/somefile', 6, 10);
    this.mockConsole.expects('log').once().withExactArgs('%s (%s)', 'some/path/somefile', 'some error');
    function on(event, cb) {
      if (event === 'uploaded') {
        cb({
          filename: 'some/path/somefile',
          transferredFileCount: 6,
          totalFileCount: 10
        });
      } else if (event === 'upload-error') {
        cb({
          relativePath: 'some/path',
          filename: 'somefile',
          err: 'some error'
        });
      }
    }
    function deploy(opts, cb) {
      assert.equals(opts.protocol, undefined);
      assert.equals(opts.include, undefined);
      assert.equals(opts.local, undefined);
      assert.equals(opts.remote, undefined);
      assert.equals(opts.host, 'somehost');
      assert.equals(opts.port, 21);
      assert.equals(opts.parallelUploads, 10);
      assert.equals(opts.localRoot, 'somelocalpath');
      assert.equals(opts.remoteRoot, 'someremotepath');
      cb();
    }

    var mockFtpDeploy = function () {
      return { on: on, deploy: deploy };
    };
    var Sendman = proxyquire('../lib/sendman', { 'ftp-deploy': mockFtpDeploy });
    var sendman = new Sendman();

    var opts = { protocol: 'ftp', host: 'somehost', local: 'somelocalpath', remote: 'someremotepath' };
    sendman._ftp(opts, done);
  }
});

buster.testCase('sendman - s3', {
  setUp: function () {
    this.mockConsole = this.mock(console);
    this.mockS3 = this.mock(s3);
  },
  'should upload objects': function (done) {
    this.mockConsole.expects('log').once().withExactArgs('Uploading %d/%d...', 23, 88);
    this.mockConsole.expects('log').once().withExactArgs('Deleting %d/%d...', 3, 10);
    this.mockConsole.expects('log').once().withExactArgs('Uploaded %d/%d', 23, 88);
    this.mockConsole.expects('log').once().withExactArgs('Deleted %d/%d', 3, 10);
    var mockUploader = {
      progressAmount: 23,
      progressTotal: 88,
      deleteAmount: 3,
      deleteTotal: 10,
      doneFindingFiles: false,
      doneFindingObjects: false,
      doneMd5: false,
      on: function (event, cb) {
        if (event === 'progress') {
          cb();
        } else if (event === 'end') {
          cb();
        }
      }
    };
    var mockClient = {
      uploadDir: function (params) {
        assert.equals(params.deleteRemoved, true);
        assert.equals(params.localDir, 'somelocalpath');
        assert.equals(params.s3Params.Bucket, 'somebucket');
        assert.equals(params.s3Params.Prefix, 'someremotepath');
        return mockUploader;
      }
    };
    this.mockS3.expects('createClient').once().withArgs({ s3Options: { accessKeyId: 'someaccesskeyid', secretAccessKey: 'somesecretaccesskey', region: 'ap-southeast-2' }}).returns(mockClient);
    var opts = { bucket: 'somebucket', local: 'somelocalpath', remote: 'someremotepath', accessKeyId: 'someaccesskeyid', secretAccessKey: 'somesecretaccesskey', region: 'ap-southeast-2' };
    var sendman = new Sendman();
    sendman._s3(opts, done);
  },
  'should display expected messages at the start of the progress': function (done) {
    this.mockConsole.expects('log').once().withExactArgs('Found %d files', 88);
    this.mockConsole.expects('log').once().withExactArgs('Done finding objects');
    this.mockConsole.expects('log').once().withExactArgs('Done MD5 hashing');
    this.mockConsole.expects('log').once().withExactArgs('Uploaded %d/%d', 0, 88);
    this.mockConsole.expects('log').once().withExactArgs('Deleted %d/%d', 0, 10);
    var mockUploader = {
      filesFound: 88,
      progressAmount: 0,
      progressTotal: 88,
      deleteAmount: 0,
      deleteTotal: 10,
      doneFindingFiles: true,
      doneFindingObjects: true,
      doneMd5: true,
      on: function (event, cb) {
        if (event === 'progress') {
          cb();
        } else if (event === 'end') {
          cb();
        }
      }
    };
    var mockClient = {
      uploadDir: function (params) {
        assert.equals(params.deleteRemoved, true);
        assert.equals(params.localDir, 'somelocalpath');
        assert.equals(params.s3Params.Bucket, 'somebucket');
        assert.equals(params.s3Params.Prefix, 'someremotepath');
        return mockUploader;
      }
    };
    this.mockS3.expects('createClient').once().withArgs({ s3Options: { accessKeyId: 'someaccesskeyid', secretAccessKey: 'somesecretaccesskey', region: 'ap-southeast-2' }}).returns(mockClient);
    var opts = { bucket: 'somebucket', local: 'somelocalpath', remote: 'someremotepath', accessKeyId: 'someaccesskeyid', secretAccessKey: 'somesecretaccesskey', region: 'ap-southeast-2' };
    var sendman = new Sendman();
    sendman._s3(opts, done);
  }
});

buster.testCase('sendman - scp', {
  setUp: function () {
    this.mockScp2 = this.mock(scp2);
  },
  'should set opts for scp module': function (done) {
    this.mockScp2.expects('scp').once().withArgs('somelocalpath', 'someusername:somepassword@somehost:someremotepath').callsArgWith(2);
    var opts = { protocol: 'scp', host: 'somehost', local: 'somelocalpath', remote: 'someremotepath', username: 'someusername', password: 'somepassword' };
    var sendman = new Sendman();
    sendman._scp(opts, done);
  }
});

buster.testCase('sendman - rsync', {
  setUp: function () {
    this.mockConsole = this.mock(console);
    this.mockRsyncwrapper = this.mock(rsyncwrapper);
  },
  'should set opts for rsync module with source and destination properties': function (done) {
    this.mockConsole.expects('log').once().withExactArgs('rsync some/source/path some/destination/path');
    this.mockConsole.expects('log').once().withExactArgs('somestdout');
    var opts = { protocol: 'rsync', source: 'some/source/path', destination: 'some/destination/path' };
    var mockRsyncwrapper = function (opts, cb) {
      assert.equals(opts, { src: 'some/source/path', dest: 'some/destination/path', recursive: true });
      cb(null, 'somestdout', null, 'rsync some/source/path some/destination/path');
    };
    var Sendman = proxyquire('../lib/sendman', { 'rsyncwrapper': mockRsyncwrapper });
    var sendman = new Sendman();
    sendman._rsync(opts, done);
  },
  'should set opts for rsync module with local and remote properties': function (done) {
    this.mockConsole.expects('log').once().withExactArgs('rsync somelocalpath someusername:somepassword@somehost:someremotepath');
    this.mockConsole.expects('error').once().withExactArgs('somestderr');
    var opts = { protocol: 'rsync', host: 'somehost', local: 'somelocalpath', remote: 'someremotepath', username: 'someusername', password: 'somepassword' };
    var mockRsyncwrapper = function (opts, cb) {
      assert.equals(opts, { src: 'somelocalpath', dest: 'someusername:somepassword@somehost:someremotepath', recursive: true });
      cb(null, null, 'somestderr', 'rsync somelocalpath someusername:somepassword@somehost:someremotepath');
    };
    var Sendman = proxyquire('../lib/sendman', { 'rsyncwrapper': mockRsyncwrapper });
    var sendman = new Sendman();
    sendman._rsync(opts, done);
  }
});


complexity/plato/report.js
__report = {"summary":{"total":{"jshint":0,"sloc":220,"maintainability":141.10565400616903},"average":{"sloc":110,"maintainability":"70.55","jshint":"0.00"}},"reports":[{"info":{"file":"lib/cli.js","fileShort":"lib/cli.js","fileSafe":"lib_cli_js","link":"files/lib_cli_js/index.html"},"jshint":{"messages":0},"complexity":{"aggregate":{"sloc":{"logical":19,"physical":32},"cyclomatic":3,"halstead":{"operators":{"distinct":10,"total":44,"identifiers":["__stripped__"]},"operands":{"distinct":28,"total":52,"identifiers":["__stripped__"]},"length":96,"vocabulary":38,"difficulty":9.285714285714286,"volume":503.8010412905842,"effort":4678.1525262697105,"bugs":0.16793368043019472,"time":259.8973625705395},"params":2,"line":1,"cyclomaticDensity":15.789473684210526,"complexity":{"cyclomatic":3,"sloc":{"logical":19,"physical":32},"halstead":{"operators":{"distinct":10,"total":44,"identifiers":["__stripped__"]},"operands":{"distinct":28,"total":52,"identifiers":["__stripped__"]},"length":96,"vocabulary":38,"difficulty":9.285714285714286,"volume":503.8010412905842,"effort":4678.1525262697105,"bugs":0.16793368043019472,"time":259.8973625705395}}},"module":"lib/cli.js","maintainability":75.02069872093183}},{"info":{"file":"lib/sendman.js","fileShort":"lib/sendman.js","fileSafe":"lib_sendman_js","link":"files/lib_sendman_js/index.html"},"jshint":{"messages":0},"complexity":{"aggregate":{"sloc":{"logical":79,"physical":188},"cyclomatic":9,"halstead":{"operators":{"distinct":17,"total":223,"identifiers":["__stripped__"]},"operands":{"distinct":97,"total":285,"identifiers":["__stripped__"]},"length":508,"vocabulary":114,"difficulty":24.974226804123713,"volume":3471.108127195689,"effort":86688.24163022224,"bugs":1.157036042398563,"time":4816.013423901236},"params":15,"line":1,"cyclomaticDensity":11.39240506329114,"complexity":{"cyclomatic":9,"sloc":{"logical":79,"physical":188},"halstead":{"operators":{"distinct":17,"total":223,"identifiers":["__stripped__"]},"operands":{"distinct":97,"total":285,"identifiers":["__stripped__"]},"length":508,"vocabulary":114,"difficulty":24.974226804123713,"volume":3471.108127195689,"effort":86688.24163022224,"bugs":1.157036042398563,"time":4816.013423901236}}},"module":"lib/sendman.js","maintainability":66.0849552852372}}]}

lib/sendman.js
var cli = require('bagofcli');
var fsx = require('fs.extra');
var FtpDeploy = require('ftp-deploy');
var p = require('path');
var rsyncwrapper = require('rsyncwrapper');
var s3 = require('s3');
var scp2 = require('scp2');
var util = require('util');

/**
 * Class Sendman
 */
function Sendman() {
}

/**
 * Create a sample .sendman.json setup file in current working directory.
 *
 * @param {Function} cb: standard cb(err, result) callback
 */
Sendman.prototype.init = function (cb) {
  fsx.copy(p.join(__dirname, '../examples/.sendman.json'), '.sendman.json', cb);
};

/**
 * Send files with properties in a configuration file.
 *
 * @param {String} file: configuration file, default: .sendman.json
 * @param {Function} cb: standard cb(err, result) callback
 */
Sendman.prototype.send = function (file, cb) {

  const DEFAULT_PORT = 21;
  const MAX_UPLOADS = 10;

  var opts = JSON.parse(cli.lookupFile(file));

  // merge sub-configuration file
  if (opts.include) {
    var moreOpts = JSON.parse(cli.lookupFile(opts.include));
    Object.keys(moreOpts).forEach(function (key) {
      opts[key] = moreOpts[key];
    });
  }

  if (['ftp', 'scp', 'rsync', 's3'].indexOf(opts.protocol) !== -1) {
    this['_' + opts.protocol](opts, cb);
  } else {
    cb(new Error('Unsupported protocol %s', opts.protocol));
  }
};

/**
 * Send files under a local path to a remote path via FTP.
 *
 * @param {Object} opts: optional
 *   - host: hostname
 *   - port: port number, default: 21
 *   - username: username
 *   - password: password
 *   - local: local path, which files will be deployed
 *   - remote: remote path where to send the local files to
 *   - parallel: the number of files to upload in parallel, default: 10
 *   - continue: if true then continue uploading after an error, default: true
 * @param {Function} cb: standard cb(err, result) callback
 */
Sendman.prototype._ftp = function (opts, cb) {

  const DEFAULT_PORT = 21;
  const DEFAULT_PARALLEL = 10;
  const DEFAULT_CONTINUE = true;

  // add alias properties
  opts.localRoot = opts.local;
  opts.remoteRoot = opts.remote;

  // set defaults
  opts.port = opts.port || DEFAULT_PORT;
  opts.parallelUploads = opts.parallel || DEFAULT_PARALLEL;
  opts.continueOnError = opts.continue || DEFAULT_CONTINUE;

  // remove temporary properties
  delete opts.protocol;
  delete opts.include;
  delete opts.local;
  delete opts.remote;

  // send file via FTP
  var ftpDeploy = new FtpDeploy();
  ftpDeploy.on('uploaded', function(data) {
    console.log('%s (%d/%d)', data.filename, data.transferredFileCount, data.totalFileCount);
  });
  ftpDeploy.on('upload-error', function (data) {
    console.log('%s (%s)', p.join(data.relativePath, data.filename), data.err);
  });
  ftpDeploy.deploy(opts, cb);
};

/**
 * Send files under a local path to a remote S3 bucket.
 *
 * @param {Object} opts: optional
 *   - local: local path, which contains files to be deployed to S3
 *   - bucket: S3 bucket name
 *   - remote: S3 bucket prefix path
 *   - region: S3 bucket region
 *   - accessKeyId: AWS access key ID
 *   - secretAccessKey: AWS secret access key
 * @param {Function} cb: standard cb(err, result) callback
 */
Sendman.prototype._s3 = function (opts, cb) {

  var client = s3.createClient({
    s3Options: {
      accessKeyId: opts.accessKeyId,
      secretAccessKey: opts.secretAccessKey,
      region: opts.region
    }
  });

  var params = {
    deleteRemoved: true,
    localDir: opts.local,
    s3Params: {
      Bucket: opts.bucket,
      Prefix: opts.remote,
    },
  };

  var uploader = client.uploadDir(params);
  var doneFindingFiles;
  var doneFindingObjects;
  var doneMd5;
  var progressCounter = 0;

  uploader.on('progress', function () {

    // log initial statuses
    if (!doneFindingFiles && uploader.doneFindingFiles === true) {
      console.log('Found %d files', uploader.filesFound);
      doneFindingFiles = true;
    }
    if (!doneFindingObjects && uploader.doneFindingObjects === true) {
      console.log('Done finding objects');
      doneFindingObjects = true;
    }
    if (!doneMd5 && uploader.doneMd5 === true) {
      console.log('Done MD5 hashing');
      doneMd5 = true;
    }

    // log upload status
    if (progressCounter % 500 === 0 && uploader.progressAmount > 0) {
      console.log('Uploading %d/%d...', uploader.progressAmount, uploader.progressTotal);
    }

    // log delete status
    if (progressCounter % 100 === 0 && uploader.deleteAmount > 0) {
      console.log('Deleting %d/%d...', uploader.deleteAmount, uploader.deleteTotal);
    }
  });
  uploader.on('error', cb);
  uploader.on('end', function () {
    console.log('Uploaded %d/%d', uploader.progressAmount, uploader.progressTotal);
    console.log('Deleted %d/%d', uploader.deleteAmount, uploader.deleteTotal);
    cb();
  });
};

/**
 * Send files under a local path to a remote path via SCP.
 *
 * @param {Object} opts: optional
 *   - host: hostname
 *   - port: port number, default: 22
 *   - username: username
 *   - password: password
 *   - local: local path, which files will be deployed
 *   - remote: remote path where to send the local files to
 * @param {Function} cb: standard cb(err, result) callback
 */
Sendman.prototype._scp = function (opts, cb) {

  const DEFAULT_PORT = 22;

  // add alias properties
  opts.localRoot = opts.local;
  opts.remoteRoot = opts.remote;

  // set defaults
  opts.port = opts.port || DEFAULT_PORT;

  // remove temporary properties
  delete opts.protocol;
  delete opts.include;
  delete opts.local;
  delete opts.remote;

  // send file via SCP
  var remoteUrl = util.format('%s:%s@%s:%s', opts.username, opts.password, opts.host, opts.remoteRoot);
  scp2.scp(opts.localRoot, remoteUrl, cb);
};

/**
 * Send files under a local path to another local or remote path via rsync.
 * Option source or destination can either be a /path/to/file or <username>@<host>:/path/to/file
 *
 * @param {Object} opts: optional
 *   - host: hostname
 *   - username: username
 *   - password: password
 *   - local: local path, which files will be deployed
 *   - remote: remote path where to send the local files to
 *   - source: local or remote path of source file/directory to be deployed
 *   - destination: local or remote path of destination file/directory to send to
 * @param {Function} cb: standard cb(err, result) callback
 */
Sendman.prototype._rsync = function (opts, cb) {

  // use remote URL construction, handy for easy switching from SCP to rsync protocol
  if (opts.local) {

    opts.src = opts.local;
    opts.dest = util.format('%s:%s@%s:%s', opts.username, opts.password, opts.host, opts.remote);

    // remove temporary properties
    delete opts.protocol;
    delete opts.include;
    delete opts.local;
    delete opts.remote;
    delete opts.host;
    delete opts.username;
    delete opts.password;

  // use freeform source and destination locations
  } else {

    opts.src = opts.source;
    opts.dest = opts.destination;

    // remove temporary properties
    delete opts.protocol;
    delete opts.source;
    delete opts.destination;

  }

  opts.recursive = true;

  rsyncwrapper(opts, function (err, stdout, stderr, cmd) {
    console.log(cmd);
    if (stdout) {
      console.log(stdout);
    }
    if (stderr) {
      console.error(stderr);
    }
    var result = {
      stdout: stdout,
      stderr: stderr,
      cmd: cmd
    };
    cb(err, result);
  });
};

module.exports = Sendman;
